### PR TITLE
Fix validation of abstract methods in decorators

### LINKED ISSUE #51196: ArC: imprecise check for abstract methods in decorators
### Describe the bug

The check for `abstract` methods declared in decorators, as present in `Decorators.createDecorator()`, is imprecise. We check that decorators do not declare _any_ `abstract` methods, but the spec says that decorators may not declare `abstract` methods that do not belong to a decorated type. Also, we only check methods declared on the decorator class itself, not inherited methods. (We can probably ignore `abstract` methods inherited from interfaces, because those are decorated types, but a decorator can inherit `abstract` methods from a superclass, and those do not necessarily have to be present on decorated types.)

That doesn't cause any TCK failures, because decorators are not present in CDI Lite (plus I'm not sure this is actually tested by the TCK), and so far there hasn't been any issues filed about this, so this is purely of a theoretical concern (I only stumbled upon this by writing test cases for various corner cases). Still, it is a deviation from the specification, so I believe we should fix it. It's a lower priority kind of thing, though.

### Expected behavior

Only fail when an `abstract` method that doesn't belong to a decorated type is explicitly declared on a decorator or inherited (and not implemented!) from a supertype.

### Actual behavior

Fail whenever an `abstract` method is explicitly declared on a decorator; `abstract` methods declared on supertypes are ignored.

### How to Reproduce?

N/A

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by fabriciolima**
This change improves the validation logic for abstract decorators to ensure that any abstract method declared in the decorator is present in the decorated types, as per the CDI specification.

Fixes #51196

### CODE REVIEW

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Decorators.java [hunk 1]
```diff
@@ -111,32 +115,89 @@ static DecoratorInfo createDecorator(ClassInfo decoratorClass, BeanDeployment be
...
+                String methods = decoratorAbstractMethods.stream()
+                        .map(MethodInfo::toString)
+                        .collect(Collectors.joining(", "));
+
+                String message = String.format(
+                        "Abstract decorator %s declares abstract methods not present on decorated interfaces: %s",
+                        decoratorClass.name(), methods);
+
+                LOGGER.warn(message);
+                throw new DefinitionException(message);
```
Comment (manovotn) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Decorators.java]: Why both?
Just the exception should be enough, they both carry the same message.

File: independent-projects/arc/processor/src/test/java/io/quarkus/arc/processor/AbstractDecoratorValidationTest.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import jakarta.enterprise.inject.spi.DefinitionException;
+import jakarta.inject.Inject;
+
+import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.Index;
+import org.jboss.jandex.IndexView;
+import org.junit.jupiter.api.Test;
+
+public class AbstractDecoratorValidationTest {
```
Comment (manovotn) [inline independent-projects/arc/processor/src/test/java/io/quarkus/arc/processor/AbstractDecoratorValidationTest.java]: I'd place these tests under `independent-projects/arc/tests/...` instead and follow the structure we have there.
See for instance [this test](https://github.com/quarkusio/quarkus/blob/main/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/generics/GenericsDecoratorTest.java#L21-L31) - the bulk of decorator tests are already there and you don't need to deal with creating an index, operating on Jandex classes or manually creating `BeanDeployment`.

**Review by fabriciolima — COMMENTED** [review]
Removing Logger warning

File: independent-projects/arc/processor/src/test/java/io/quarkus/arc/processor/AbstractDecoratorValidationTest.java (hunk 2 above)
Comment (fabriciolima) [reply independent-projects/arc/processor/src/test/java/io/quarkus/arc/processor/AbstractDecoratorValidationTest.java]: I have see this test(made a small change there too) but where i work we have to make a test for each issue.
Case you think better(as do i) to put all in GenericsDecoratorTest.java i can do that

Comment (manovotn) [reply independent-projects/arc/processor/src/test/java/io/quarkus/arc/processor/AbstractDecoratorValidationTest.java]: I'd create a separate class but please do that under `independent-projects/arc/tests/...`.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Decorators.java [hunk 3]
```diff
@@ -111,32 +115,88 @@ static DecoratorInfo createDecorator(ClassInfo decoratorClass, BeanDeployment be
...
+    private static boolean compare(MethodInfo method, Set<MethodInfo> interfaceMethods) {
+        return interfaceMethods.stream()
+                .filter(ifaceMethod -> ifaceMethod.name().equals(method.name()))
+                .filter(ifaceMethod -> ifaceMethod.parameterTypes().size() == method.parameterTypes().size())
+                .anyMatch(ifaceMethod -> IntStream.range(0, method.parameterTypes().size())
+                        .allMatch(i -> {
+                            Type ifaceParam = ifaceMethod.parameterTypes().get(i);
+                            Type implParam = method.parameterTypes().get(i);
+                            return ifaceParam.name().equals(DotNames.OBJECT) ||
+                                    ifaceParam.name().equals(implParam.name());
```
Comment (manovotn) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Decorators.java:199]: I didn't try to write a test for this yet (sorry, will have more time for this after Christmas!) but my guess is that this won't be sufficient for generic decorator class hierarchy and we will need parameter substitution.
Aka similar to what @Ladicek said in https://github.com/quarkusio/quarkus/pull/51439#discussion_r2597430532

File: independent-projects/arc/processor/src/test/java/io/quarkus/arc/processor/AbstractDecoratorValidationTest.java (hunk 2 above)
Comment (fabriciolima) [reply independent-projects/arc/processor/src/test/java/io/quarkus/arc/processor/AbstractDecoratorValidationTest.java]: > I'd create a separate class but please do that under `independent-projects/arc/tests/...`.

put everything in independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/validation/AbstractDecoratorWithAbstractMethodTest.java

**Comment by geoand** [comment]
@manovotn can you please take another look at this?

**Comment by Ladicek** [comment]
I started rewriting this, because the validation is incomplete. I should have it done tomorrow, probably.

**Comment by geoand** [comment]
Oh nice, thanks for the update

**Comment by manovotn** [comment]
I have it on my list but there were other items taking precedence so far.

Thanks @Ladicek!

**Comment by Ladicek** [comment]
Submitted #51893.
